internal/stores/contracts.go

```diff
@@ -179,3 +194,382 @@ func NewJSONContractStore(dir string) (*JSONContractStore, error) {
+	if len(c.Signatures) != len(signatures) {
+		return rhpv2.Contract{}, fmt.Errorf("contract in db got %v signatures but expected %v", len(c.Signatures), len(signatures))
+	}
+	for i, sig := range c.Signatures {
+		signatures[i] = types.TransactionSignature{
+			ParentID:       crypto.Hash(sig.ParentID),
+			PublicKeyIndex: sig.PublicKeyIndex,
+			Timelock:       sig.Timelock,
+			CoveredFields:  sig.CoveredFields,
+			Signature:      sig.Signature,
+		}
+	}
+
+	return rhpv2.Contract{
+		Revision:   revision,
+		Signatures: signatures,
+	}, nil
+}
+
+// NewSQLContractStore creates a new SQLContractStore from a given gorm
+// Dialector.
+func NewSQLContractStore(conn gorm.Dialector, migrate bool) (*SQLContractStore, error) {
+	db, err := gorm.Open(conn, &gorm.Config{})
+	if err != nil {
+		return nil, err
+	}
+
+	if migrate {
+		// Create the tables.
+		tables := []interface{}{
+			&dbContractRHPv2{},
+			&dbFileContractRevision{},
+			&dbTransactionSignature{},
+			&dbUnlockConditions{},
+			&dbSiaPublicKey{},
+			&dbValidSiacoinOutput{},
+			&dbMissedSiacoinOutput{},
+			&dbHostSet{},
+			&dbHostSetEntry{},
+		}
+		if err := db.AutoMigrate(tables...); err != nil {
+			return nil, err
+		}
+		if res := db.Exec("PRAGMA foreign_keys = ON", nil); res.Error != nil {
+			return nil, res.Error
+		}
+	}
+
+	return &SQLContractStore{
+		db: db,
+	}, nil
+}
+
+// AddContract implements the bus.ContractStore interface.
+func (s *SQLContractStore) AddContract(c rhpv2.Contract) error {
+	return s.db.Transaction(func(tx *gorm.DB) error {
+		fcid := c.ID()
+
+		// Prepare valid and missed outputs.
+		newValidOutputs := make([]dbValidSiacoinOutput, 0, len(c.Revision.NewValidProofOutputs))
+		for _, sco := range c.Revision.NewValidProofOutputs {
+			newValidOutputs = append(newValidOutputs, dbValidSiacoinOutput{
+				ParentID:   fcid,
+				UnlockHash: sco.UnlockHash,
+				Value:      sco.Value.Big(),
+			})
+		}
+		newMissedOutputs := make([]dbMissedSiacoinOutput, 0, len(c.Revision.NewMissedProofOutputs))
+		for _, sco := range c.Revision.NewMissedProofOutputs {
+			newMissedOutputs = append(newMissedOutputs, dbMissedSiacoinOutput{
+				ParentID:   fcid,
+				UnlockHash: sco.UnlockHash,
+				Value:      sco.Value.Big(),
+			})
+		}
+
+		// Prepare pubkeys for unlock conditions.
+		pubkeys := make([]dbSiaPublicKey, 0, len(c.Revision.UnlockConditions.PublicKeys))
+		for _, pk := range c.Revision.UnlockConditions.PublicKeys {
+			pubkeys = append(pubkeys, dbSiaPublicKey{
+				Algorithm:         pk.Algorithm,
+				Key:               pk.Key,
+				UnlockConditionID: fcid,
+			})
+		}
+
+		// Prepare unlock conditions.
+		unlockConditions := dbUnlockConditions{
+			ParentID:           fcid,
+			Timelock:           c.Revision.UnlockConditions.Timelock,
+			PublicKeys:         pubkeys,
+			SignaturesRequired: c.Revision.UnlockConditions.SignaturesRequired,
+		}
+
+		// Prepare contract revision.
+		revision := dbFileContractRevision{
+			//ParentID: fcid,
```

Comment:
oversight? kind of weird because your tests assert a working cascade delete

Reply:
Hm it might just do some magic due to the foreign key constraint